### PR TITLE
changed DataLoader

### PR DESCRIPTION
**Pull request by spirinamayya**
Changed DataLoader, added comments

### CODE REVIEW

File: rectools/models/sasrec.py [hunk 1]
```diff
@@ -51,7 +50,7 @@ def __init__(
         self.item_net_dropout_rate = item_net_dropout_rate
         self.use_pos_emb = use_pos_emb
         self.model: SASRec
-        self.task_converter = SequenceTaskConverter(self.session_maxlen)
+        self.task_converter = SequenceTaskConverter()
```
Comment (blondered) [inline rectools/models/sasrec.py]: `SequenceTaskConverter` - это класс, у которого нет конструктора и есть всего один метод. Это верный сигнал к тому, что это должен быть не класс, а функция. 
Остально решить, должна ли это быть отдельный функция или метод другого класса (или просто часть кода в другом методе). Из кандидатов: `SequenceDataset` либо `SasRecRecommenderModel`. Давай решать.

Мне кажется, что мы в целом можем реализовать у `SequenceDataset` метод (classmethod) `from_dataset`, который будет принимать rectools.Dataset и там прогонять все манипуляции с данными. По аналогии с https://github.com/spirinamayya/RecTools/blob/cb8d32987b476846761867b164e2041ddd8448e4/rectools/dataset/torch_datasets.py#L81
Тогда  мы перетащим в метод `from_dataset` весь код начиная с https://github.com/spirinamayya/RecTools/blob/cb8d32987b476846761867b164e2041ddd8448e4/rectools/models/sasrec.py#L71

File: rectools/models/sasrec.py [hunk 2]
```diff
@@ -87,13 +86,17 @@ def _fit(
...
         logger.info("converting datasets to task format")
-        train_x, train_y = self.task_converter.train_transform(
+        train_x = self.task_converter.transform(
             user_item_interactions=user_item_interactions,
         )
 
         logger.info("building train dataset")
-        train_dataset = SequenceDataset(x=train_x, y=train_y, batch_size=self.batch_size)
-        train_dataloader = DataLoader(train_dataset, batch_size=None)  # batching in dataset
+        train_dataset = SequenceDataset(x=train_x, session_maxlen=self.session_maxlen)
```
Comment (blondered) [inline rectools/models/sasrec.py]: вот здесь будем вызывать `SequenceDataset.from_dataset(dataset, self.session_maxlen)`

File: rectools/models/sasrec.py [hunk 3]
```diff
@@ -144,68 +147,24 @@ def recommend(
 
 @dataclass
 class SASRecModelInput:
-    """sessions (torch.LongTensor): [batch_size, maxlen] user history"""
-
-    sessions: torch.LongTensor
-
+    """
```
Comment (blondered) [inline rectools/models/sasrec.py]: Дропаем. SequenceDataset будет выдавать Tuple

File: rectools/models/sasrec.py [hunk 4]
```diff
@@ -267,35 +187,70 @@ class SequenceDataset(Dataset):
...
+    sessions: List[List[int]],  # [batch_size, session_len]
+    weights: List[List[float]],  # [batch_size, session_len]
+    session_maxlen: int,
+) -> Tuple[torch.LongTensor, torch.LongTensor, torch.LongTensor]:
+    """TODO"""
+    x = np.zeros((len(sessions), session_maxlen))  # [batch_size, session_maxlen]
+    y = np.zeros((len(sessions), session_maxlen))  # [batch_size, session_maxlen]
+    yw = np.zeros((len(sessions), session_maxlen))  # [batch_size, session_maxlen]
+    # left padding
+    for i, (ses, ses_weights) in enumerate(zip(sessions, weights)):
```
Comment (blondered) [inline rectools/models/sasrec.py:192]: Итерация по строкам обычно сильно медленнее, чем векторные операции. Давай попробуем альтернативные реализации и замерим время в тестовом ноутбуке - что отработает быстрее?

Предлагаю воспользоваться torch.nn.utils.rnn.pad_sequence. Проблема только в том, что эта функция будет выдавать максимальную длину вектора в батче, а не фиксированную session_maxlen. Это можно обойти, добавив (а потом удалив) строчку с нулями, например. Вот тут есть пример:
https://discuss.pytorch.org/t/how-to-create-batches-of-a-list-of-varying-dimension-tensors/50773/16#:~:text=def-,collate_fn_padd,-(batch)%3A%0A%20%20%20%20%27%27%27%0A%20%20%20%20Padds%20batch

После того как решим с паддингом, надо будет ещё посмотреть, как нам сделать to(device) в collate_fn перед паддингом, это тоже должно нас чуть ускорить

**Review by blondered — CHANGES_REQUESTED** [review]

File: rectools/models/sasrec.py [hunk 5]
```diff
@@ -312,8 +267,10 @@ def __init__(self, factors: int, dropout_rate: float):
 
     def forward(self, inputs: torch.Tensor) -> torch.Tensor:
         """TODO"""
+        # inputs: [batch_size, factors, session_maxlen] -> outputs: [batch_size, session_maxlen, factors]
+        # [batch_size, factors, session_maxlen]
         outputs = self.dropout2(self.conv2(self.relu(self.dropout1(self.conv1(inputs.transpose(-1, -2))))))
-        outputs = outputs.transpose(-1, -2)  # as Conv1D requires (N, C, Length)
+        outputs = outputs.transpose(-1, -2)  # as Conv1D requires (N, C, Length) [batch_size, session_maxlen, factors]
         outputs += inputs
```
Comment (blondered) [inline rectools/models/sasrec.py:236]: Если inputs и outputs имеют разные размерности, то как мы их складываем в skip connection?

File: rectools/models/sasrec.py [hunk 6]
```diff
@@ -267,35 +187,70 @@ class SequenceDataset(Dataset):
...
+        # session_len = len(session)
+        return session, weights
+
+
+def collate_fn_train(
+    batch: List[Tuple[List[int], List[float]]],
+    session_maxlen: int,
+) -> Tuple[torch.LongTensor, torch.LongTensor, torch.LongTensor]:
+    """TODO"""
+    sessions = list(list(zip(*batch))[0])  # [batch_size, session_len]
```
Comment (blondered) [inline rectools/models/sasrec.py:185]: а простое `sessions, weights = batch` здесь не отработает?

и всю логику из sessions_transform_train можно перенести сюда, потому что функции делают общее дело без необходимости их разделять. с рекомменд аналогично

File: rectools/models/sasrec.py [hunk 7]
```diff
@@ -341,7 +298,7 @@ def forward(self) -> torch.Tensor:
 
     def get_all_embeddings(self) -> torch.Tensor:
         """TODO"""
-        item_embs = self.item_emb(self.catalogue)
+        item_embs = self.item_emb(self.catalogue)  # [n_items, factors]
```
Comment (blondered) [inline rectools/models/sasrec.py]:
```suggestion
        item_embs = self.item_emb(self.catalogue)  # [n_items + 1, factors]
```

File: rectools/models/sasrec.py [hunk 8]
```diff
@@ -577,18 +537,17 @@ def fit(
     def train_step(self, x: torch.Tensor, y: torch.Tensor, w: torch.Tensor) -> None:
         """TODO"""
         self.optimizer.zero_grad()
-        logits = self.model(x)
+        logits = self.model(x)  # [batch_size, session_maxlen, n_items + 1]
         # We are using CrossEntropyLoss with a multi-dimensional case
 
-        # Logits must be passed in form of [batch_size, n_classes, session_position],
-        # where session_position is one extra dimension
+        # Logits must be passed in form of [batch_size, session_maxlen, n_items + 1],
```
Comment (blondered) [inline rectools/models/sasrec.py]:
```suggestion
        # Logits must be passed in form of [batch_size, n_items + 1, session_maxlen] where n_items + 1 is number of classes
```

File: rectools/models/sasrec.py [hunk 9]
```diff
@@ -451,43 +409,46 @@ def log2feats(self, sessions: torch.Tensor, item_embs: torch.Tensor) -> torch.Te
             torch.Tensor. [batch_size, history_len, emdedding_dim]
 
         """
-        seqs = item_embs[sessions]
+        seqs = item_embs[sessions]  # [batch_size, session_maxlen, factors]
 
         if self.use_pos_emb:
-            positions = np.tile(np.array(range(sessions.shape[1])), [sessions.shape[0], 1])
+            positions = np.tile(
```
Comment (blondered) [inline rectools/models/sasrec.py]: Давай ещё добавим коммент # TODO: add inverse positional embedding (much talked about with good feedback)

File: rectools/models/sasrec.py [hunk 10]
```diff
@@ -451,43 +409,46 @@ def log2feats(self, sessions: torch.Tensor, item_embs: torch.Tensor) -> torch.Te
...
 
         seqs = self.encoder(
             seqs=seqs,
             attention_mask=attention_mask,
             timeline_mask=timeline_mask,
-        )
+        )  # [batch_size, session_maxlen, factors]
 
-        log_feats = self.last_layernorm(seqs)  # (U, T, C) -> (U, -1, C)
+        log_feats = self.last_layernorm(seqs)  # (U, T, C) -> (U, -1, C) [batch_size, session_maxlen, factors]
```
Comment (blondered) [inline rectools/models/sasrec.py]: Давай дропнем  `(U, T, C) -> (U, -1, C)`. Я не очень понимаю, про что это

File: rectools/models/sasrec.py [hunk 11]
```diff
@@ -451,43 +409,46 @@ def log2feats(self, sessions: torch.Tensor, item_embs: torch.Tensor) -> torch.Te
...
 
         seqs = self.emb_dropout(seqs)
 
-        timeline_mask = sessions == 0
-        seqs *= ~timeline_mask.unsqueeze(-1)  # broadcast in last dim
+        timeline_mask = sessions == 0  # [batch_size, session_maxlen]
+        seqs *= ~timeline_mask.unsqueeze(-1)  # broadcast in last dim [batch_size, session_maxlen, factors]
 
         # TODO do we need to mask padding embs attention?
         tl = seqs.shape[1]  # time dim len for enforce causality
```
Comment (blondered) [inline rectools/models/sasrec.py]: Мы несколько раз в коде это значение по-разному достаём. Вообще говоря это session_maxlen. Но ок, можно на всякий добавить себе гибкости и вытаскивать его из батча, не обращаясь в self.session_maxlen.
Тогда в начале метода давай сразу достанем `session_maxlen = sessions.shape[1]` и будем использовать его вместо tl и в строчке, где мы positions генерим

коммент можно убрать: # time dim len for enforce causality

File: rectools/models/sasrec.py [hunk 12]
```diff
@@ -393,11 +350,11 @@ def forward(
     ) -> torch.Tensor:
         """TODO"""
         for i, _ in enumerate(self.attention_layers):
-            seqs = torch.transpose(seqs, 0, 1)
-            q = self.attention_layernorms[i](seqs)
+            seqs = torch.transpose(seqs, 0, 1)  # [session_maxlen, batch_size, factors]
```
Comment (blondered) [inline rectools/models/sasrec.py]: Давай попробуем убрать здесь двойной transpose туда-обратно.
В отдельном коммите (надо в ноутбук проверить, что мы не повлияли на качество модели): 
в `MultiheadAttention` передаём параметр `batch_first=True`, а здесь убираем transpose на строчках 353 и 357

File: rectools/models/sasrec.py [hunk 13]
```diff
@@ -451,43 +409,46 @@ def log2feats(self, sessions: torch.Tensor, item_embs: torch.Tensor) -> torch.Te
...
+            positions = np.tile(
+                np.array(range(sessions.shape[1])), [sessions.shape[0], 1]
+            )  # [batch_size, session_maxlen]
             seqs += self.pos_emb(torch.LongTensor(positions).to(self.get_model_device()))
 
         seqs = self.emb_dropout(seqs)
 
-        timeline_mask = sessions == 0
-        seqs *= ~timeline_mask.unsqueeze(-1)  # broadcast in last dim
+        timeline_mask = sessions == 0  # [batch_size, session_maxlen]
```
Comment (blondered) [inline rectools/models/sasrec.py]: давай сразу сделаем `timeline_mask = (sessions != 0).unsqueeze(-1)`, чтобы не дёргать это несколько раз в коде.
и перетащим до добавления pos_emb. домножать на timeline_mask перед энкодером нужно только если был использован positional эмбед (иначе там и так нули везде где нужно)

File: rectools/models/sasrec.py [hunk 14]
```diff
@@ -451,43 +409,46 @@ def log2feats(self, sessions: torch.Tensor, item_embs: torch.Tensor) -> torch.Te
...
-        sessions: torch.Tensor,
+        sessions: torch.Tensor,  # [batch_size, session_maxlen]
     ) -> torch.Tensor:
         """TODO"""
         # TODO merge item model with log2feats
-        item_embs = self.item_model.get_all_embeddings()
+        item_embs = self.item_model.get_all_embeddings()  # [n_items + 1, factors]
 
-        log_feats = self.log2feats(sessions, item_embs)
+        log_feats = self.log2feats(sessions, item_embs)  # [batch_size, session_maxlen, factors]
```
Comment (blondered) [inline rectools/models/sasrec.py]: давай назовём этот метод `encode_sessions` и сделаем `session_embs = self.encode_sessions(sessions, item_embs)`

и везде в коде всё что про log и feats аналогично переименуем

File: rectools/models/sasrec.py [hunk 15]
```diff
@@ -145,26 +138,44 @@ def recommend(
...
-@dataclass
-class SASRecModelInput:
-    """
-    Input for SequenceDataset
-    sessions: List[List[int]] # [n_users, session_len]
-    weights: List[List[float]] # [n_users, session_len]
-    """
+# TODO: think about moving all data processing to SequenceDataset.from_dataset method.
+# This would also mean creating different dataset types for fit and recommend
+
```
Comment (blondered) [inline rectools/models/sasrec.py:144]: Давай добавим еще вот что (обсудим с ребятами на финальном ревью):
`# If we have different dataset types for fit and recommend, we can also split to x/y/w and pad with zeros in dataset, not dataloader`

File: rectools/models/sasrec.py [hunk 16]
```diff
@@ -145,26 +138,44 @@ def recommend(
...
 
-    sessions: List[List[int]]
-    weights: List[List[float]]
+class SequenceDatasetBase(Dataset):
```
Comment (blondered) [inline rectools/models/sasrec.py]: А зачем он нам нужен если SequenceDataset у нас только один и в текущем варианте кода других не будет?

File: rectools/models/sasrec.py (hunk 6 above)
Comment (blondered) [reply rectools/models/sasrec.py:185]: ` sessions, weights = batch` ?

File: rectools/models/sasrec.py [hunk 17]
```diff
@@ -580,14 +553,14 @@ class SASRecRecommender:
     def __init__(
         self,
         model: SASRec,
-        task_converter: SequenceTaskConverter,
+        # task_converter: SequenceTaskConverter,
```
Comment (blondered) [inline rectools/models/sasrec.py]: Убираем комменты перед мерджем

File: rectools/models/sasrec.py [hunk 18]
```diff
@@ -409,46 +380,47 @@ def log2feats(self, sessions: torch.Tensor, item_embs: torch.Tensor) -> torch.Te
...
+        return session_embs
 
     # TODO check user_ids
     def forward(
         self,
         sessions: torch.Tensor,  # [batch_size, session_maxlen]
     ) -> torch.Tensor:
         """TODO"""
-        # TODO merge item model with log2feats
+        # TODO merge item model with encode_sessions
```
Comment (blondered) [inline rectools/models/sasrec.py]: Можно убрать этот коммент

File: rectools/models/sasrec.py [hunk 19]
```diff
@@ -471,10 +443,10 @@ def predict(
...
         """
-        item_embs = self.item_model.get_all_embeddings()  # [n_items, factors]
-        log_feats = self.log2feats(sessions, item_embs)  # [batch_size, session_maxlen, factors]
+        item_embs = self.item_model.get_all_embeddings()  # [n_items + 1, factors]
+        session_embs = self.encode_sessions(sessions, item_embs)  # [batch_size, session_maxlen, factors]
 
-        final_feat = log_feats[:, -1, :]  # only use last QKV classifier, a waste [batch_size, factors]
+        final_feat = session_embs[:, -1, :]  # only use last QKV classifier, a waste [batch_size, factors]
 
         item_embs = item_embs[item_indices]  # (U, I, C) [n_items, factors]
```
Comment (blondered) [inline rectools/models/sasrec.py]: А что такое (U, I, C)? (предлагаю дропнуть, но мне вообще просто интересно, какой за этим смысл

File: rectools/models/sasrec.py [hunk 20]
```diff
@@ -409,46 +380,47 @@ def log2feats(self, sessions: torch.Tensor, item_embs: torch.Tensor) -> torch.Te
...
 
+        # TODO: add inverse positional embedding (much talked about with good feedback)
         if self.use_pos_emb:
-            positions = np.tile(
-                np.array(range(sessions.shape[1])), [sessions.shape[0], 1]
-            )  # [batch_size, session_maxlen]
+            positions = np.tile(np.array(range(session_maxlen)), [sessions.shape[0], 1])  # [batch_size, session_maxlen]
             seqs += self.pos_emb(torch.LongTensor(positions).to(self.get_model_device()))
 
+        # TODO: can omit masking pad elements in embeddings
```
Comment (blondered) [inline rectools/models/sasrec.py]: Давай подробнее
`# TODO: do we need to fill padding embeds in sessions to all zeros or should we use the learnt padding embedding? Should we make it an option for user to decide?

File: rectools/models/sasrec.py (hunk 12 above)
Comment (blondered) [reply rectools/models/sasrec.py]: Давай возьмём вариант, где ты убрала transpose.  Метрики меняются от инициализации. Логику модели мы точно не ломаем

File: rectools/models/sasrec.py [hunk 21]
```diff
@@ -267,7 +238,7 @@ def __init__(self, factors: int, dropout_rate: float):
 
     def forward(self, inputs: torch.Tensor) -> torch.Tensor:
         """TODO"""
-        # inputs: [batch_size, factors, session_maxlen] -> outputs: [batch_size, session_maxlen, factors]
+        # inputs: [batch_size, session_maxlen, factors] -> outputs: [batch_size, session_maxlen, factors]
         # [batch_size, factors, session_maxlen]
         outputs = self.dropout2(self.conv2(self.relu(self.dropout1(self.conv1(inputs.transpose(-1, -2))))))
         outputs = outputs.transpose(-1, -2)  # as Conv1D requires (N, C, Length) [batch_size, session_maxlen, factors]
```
Comment (blondered) [inline rectools/models/sasrec.py]:
```suggestion
        inputs = inputs.transpose(-1, -2)  # [batch_size, session_maxlen, factors] -> [batch_size, factors, session_maxlen]
        outputs = self.dropout2(self.conv2(self.relu(self.dropout1(self.conv1(inputs)))))
        outputs = outputs.transpose(-1, -2)  #  [batch_size, factors, session_maxlen] -> [batch_size, session_maxlen, factors]
```

**Review by blondered — APPROVED** [review]
